include/albatross/src/cereal/suitesparse.hpp

```diff
@@ -0,0 +1,486 @@
+  ar(::cereal::make_nvp("cc.SPQR_istat5", cc.SPQR_istat[5]));
+  ar(::cereal::make_nvp("cc.SPQR_istat6", cc.SPQR_istat[6]));
+  ar(::cereal::make_nvp("cc.SPQR_istat7", cc.SPQR_istat[7]));
+  ar(::cereal::make_nvp("cc.SPQR_istat8", cc.SPQR_istat[8]));
+  ar(::cereal::make_nvp("cc.SPQR_istat9", cc.SPQR_istat[9]));
+
+  // Completely ignore all the GPU stuff.
+#ifdef GPU_BLAS
+  static_assert(false,
+                "This codec will not work for CHOLMOD / SPQR using GPU!");
+#endif  // GPU_BLAS
+  ar(CEREAL_NVP(cc.useGPU));
+  ALBATROSS_ASSERT(cc.useGPU == 0);
+
+  ar(CEREAL_NVP(cc.syrkStart));
+  ar(CEREAL_NVP(cc.cholmod_cpu_gemm_time));
+  ar(CEREAL_NVP(cc.cholmod_cpu_syrk_time));
+  ar(CEREAL_NVP(cc.cholmod_cpu_trsm_time));
+  ar(CEREAL_NVP(cc.cholmod_cpu_potrf_time));
+  ar(CEREAL_NVP(cc.cholmod_assemble_time));
+  ar(CEREAL_NVP(cc.cholmod_assemble_time2));
+  ar(CEREAL_NVP(cc.cholmod_cpu_gemm_calls));
+  ar(CEREAL_NVP(cc.cholmod_cpu_syrk_calls));
+  ar(CEREAL_NVP(cc.cholmod_cpu_trsm_calls));
+  ar(CEREAL_NVP(cc.cholmod_cpu_potrf_calls));
+}
+
+template <class Matrix>
+inline std::size_t get_element_size_bytes(const Matrix &m) {
+  return m.dtype == CHOLMOD_DOUBLE ? sizeof(double) : sizeof(float);
+}
+
+template <class Matrix>
+inline std::size_t get_num_x_elements(const Matrix &m) {
+  // Complex matrices store twice as many elements.
+  // https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/CHOLMOD/Core/cholmod_sparse.c#L209
+  return m.nzmax * (m.xtype == CHOLMOD_COMPLEX ? 2 : 1);
+}
+
+template <class Matrix>
+inline std::size_t get_p_element_size_bytes(const Matrix &m) {
+  return m.itype == CHOLMOD_INT ? sizeof(int) : sizeof(SuiteSparse_long);
+}
+
+template <class Matrix>
+inline std::size_t get_integer_size_bytes(const Matrix &m) {
+  return m.itype == CHOLMOD_LONG ? sizeof(SuiteSparse_long) : sizeof(int);
+}
+
+template <class Archive>
+inline void save(Archive &ar, cholmod_sparse const &m,
+                 std::uint32_t version ALBATROSS_UNUSED) {
+  ar(CEREAL_NVP(m.nrow));
+  ar(CEREAL_NVP(m.ncol));
+  ar(CEREAL_NVP(m.nzmax));
+  ar(CEREAL_NVP(m.stype));
+  ar(CEREAL_NVP(m.itype));
+  ar(CEREAL_NVP(m.xtype));
+  ar(CEREAL_NVP(m.dtype));
+  ar(CEREAL_NVP(m.sorted));
+  ar(CEREAL_NVP(m.packed));
+
+  const std::size_t element_size_bytes = get_element_size_bytes(m);
+  const std::size_t x_elements = get_num_x_elements(m);
+  const std::size_t p_size_bytes = get_p_element_size_bytes(m);
+  const std::size_t integer_size_bytes = get_integer_size_bytes(m);
+
+  detail::encode_array(ar, "m.p", m.p, (m.ncol + 1) * p_size_bytes);
+  detail::encode_array(ar, "m.i", m.i, m.nzmax * integer_size_bytes);
+  if (nullptr != m.nz && !m.packed) {
+    detail::encode_array(ar, "m.nz", m.nz, m.ncol * integer_size_bytes);
+  }
+  if (nullptr != m.x) {
+    detail::encode_array(ar, "m.x", m.x, x_elements * element_size_bytes);
+  }
+  // This should only be non-null for "zomplex" elements
```

Comment:
complex?

Reply:
no . . . sadly "zomplex" is an internal suitesparse representation.